### ISSUE TITLE
Latency tracking specification

### ISSUE THREAD

**Issue by conal** [AtticusKuhn/parallel-algorithms#1]
What is a precise and compelling specification for parallel computations that track latency/depth? We need such a specification in order to understand what it would mean for a latency tracking implementation to be correct.

**Comment by AtticusKuhn** [comment]
That is a very good point.

This is just my initial ideas on the matter, I haven't researched the literature on this (perhaps I should).

Define an **information dependency graph** of an algorithm as a directed graph where each node is an event in the algorithm and draw an edge from A to B if B requires information from A to compute. 

I have drawn an example of this as 
[a mermaid diagram](https://mermaid.live/edit#pako:eNpNkMtugzAQRX_FGomdIUABgReVkrDtqrvKUuSAE1yBjWCImiL-vcZEol75zOvO3BkqU0tg4HkeaRD7kR0OnRw6oeqgVQ95kLVCwvWIAmWpxH0Qnf-IuSb2ifZuBoVNd7HpAanSCpVoL-b6LSukV4NoOuL77-S4NRwdnMh_Om9wclC-UueNNigd7GJS1xTtjkrvWmh6W8w1UHitb4-a134O2MhOcmD2W8ubmFrkwPViS8WE5vOpK2A4TJLC1Nf7mcBuoh1ttBca2Aw_wOIwD6L4Lc-SNImLMEwoPIFlWZBkRZSlSZFHYZJHC4VfY-yAMEgprA6a4WPz2dntRn65ilV3-QPZe3e-
)


There are two mathematical ways I can think of for modelling this graph: as a bounded semi-lattice or as a category.

If we model this as a bounded semi-lattice, the meet of A and B is the latest event which gives information to both A and B. The join of A and B is the earliest event which depends on both A and B.  The start of the algorithm is the bottom (no information) and the end of the algorithm is the top (all the information). The algorithm, in this case, is a monotonic increase in information.

If we model this as a category, the objects are events and the arrows are information dependencies between the events. The start of the algorithm is the initial object, and the end of the algorithm is the terminal object.

Now we define a **partition** on the graph. A partition is a set of sets of events, where each event belongs to exactly one subset. If A -> B, then A and B cannot be in the same subset. My intuition for this is that the partition is an equivalence relation where two events are equivalent if they can be computed at the same time. I.e. A ~ B if A does not depend on B and B does not depend on A.

In my example, the **sequential partition** would be {{A}, {B}, {C}, {D}} and the **parallel partition** would be {{A}, {B,C}, {D}}. 

In this model, every execution strategy is a function from the information dependency graph to a partition of its events.

We define the **cost** of an execution strategy as the cardinality of the partition. In my example, the cost of the sequential execution strategy would be 4 and the cost of the parallel execution strategy would be 3. 

This is just my initial thoughts on how to define latency.

**Comment by AtticusKuhn** [comment]
Perhaps, I should note that such a parallel partition is not unique,

Consider [this example](https://mermaid.live/edit#pako:eNpVkDFvgzAQhf-KdRKbAUOMBR4qNSHdOnWrvFixkyCBjcBESRH_vcZEanrTffdOd09vhpNVGjhEUYSuzvUjT9NOD51sVNI2N51q1TgkzOik03UjL4Ps4lsuDPL1juL4De1f4bDBIUCNNqoDfaBX7fik4z9tv5EHYQDD04n3N6-6AHfVnRbAfav0WU6tEyDM4lfl5OzXw5yAu2HSGKZe_TkGfpbt6Ke9NMBnuAPPSZlk-a5ktKB5RQjF8ADOWEJZlbGCVmVGaJktGH6s9QdIUmBYw7DD5xZZSC6c_A4b69_lF7X8Xv8)

We could have the parallel partition {{A}, {B,C}, {D,E}, {F}} or we could have the partition {{A}, {C}, {B,D,E}, {F}}. Either way, both partitions have cardinality 4.

**Comment by conal** [comment]
(and thus correctness theorems)> This is just my initial ideas on the matter,

I like these ideas, and I think you’ll learn a lot from formalizing your specifications and proving implementation correctness. I’ll encourage you to look for opportunities to reduce complexity and increase naturalness of specifications—both being crucial to scalably correct engineering.

**Comment by AtticusKuhn** [comment]
I agree with you that my previous definition of runtime was insufficient, so I have been thinking more about the denotationally correct
and most elegant way of defining runtime.

I believe this definition is much more elegant. I am inspired by math. To me, if it is not inspired by math, then it is not elegant and therefore wrong. I was inspired by two concepts
1. metric spaces
2. the derivative.

Here is my new definition. I will be writing in pseudo-Agda to get the ideas across. Define a **cost function** as any function with type signature

```agda
cost : (A ⇨ B) → A → ℝ
```
that satisfies the following axioms

```agda
-- identity
cost id x = 0
-- triangle inequality
cost (f ∘ g)  x ≤ cost f (g x) + cost g x
```

a cost function is general enough to track memory cost or time cost, although I will mostly be focusing on time cost.
Define the **parallel cost** as any cost function satisfying some additional axioms

```agda
parallelCost : (A ⇨ B) → A → ℝ
parallelCost id x = 0
parallelCost (f ∘ g) x = parallelCost f (g x) + parallelCost g x
parallelCost exl x = 0
parallelCost exr x = 0
parallelCost (f ▵ g) x = parallelCost f x ⊔ parallelCost g x
```

And we also have the laws that

```agda

id ∘ f = f
f ∘ id = f
exl ▵ exr = id
```

Now if we define `reduce` as

```agda
reduce : (n : ℕ) → (tree n A ⇨ A)
reduce zero = id
reduce (suc n) = add ∘ twice (reduce n)
```

We can calculate the cost of any call, for example

```agda
parallelCost (reduce 2) n
 = parallelCost (add ∘ (add ∘ exl) ▵ (ad ∘ exr)) n
 = parallelCost add (exl n) ⊔ parallelCost add (exr n) + parallelCost add ((add ∘ exl ▵ add ∘ exr) n)
```

Now we may assume that the cost of the add function is constant, i.e. it takes the same amount of time to add any two numbers regardless of their size, which would add the axiom

```agda
parallelCost add x = m
```

This allows us to get a closed form for the runtime.


```agda

parallelCost (reduce (suc n)) x
 = parallelCost (add ∘ twice (reduce n)) x
 = parallelCost add (twice (reduce n) x) + parallelCost (reduce n) x
 = m + parallelCost (reduce n) x
```

Solving this recurrence relation gives us

```agda
parallelCost (reduce n) x = m * n
```

We can also do this same process with `sequentialCost`, in which case we need an alternate axiom.


```agda
sequentialCost (f ▵ g) x = sequentialCost f x  +  sequentialCost g x
```


This gives the equation

```agda
sequentialCost (reduce (suc n)) x
 = sequentialCost (add ∘ twice (reduce n)) x
 = sequentialCost add (twice (reduce n) x) + sequentialCost (reduce n) (exl n) + sequentialCost (reduce n) (exr n)
```

Solving this recursive equation gives us

```agda
sequentialCost (reduce n) x = (2 ^ n - 1) * m
```


Anyway, I hope these ideas are more elegant, simple, and correct in your eyes.
